src/generic/FreeAssAlgebraGroebner.jl

```diff
@@ -0,0 +1,641 @@
+        end
+    end
+    r = FreeAssAlgElem{T}(R, rcoeffs, rexps, length(rcoeffs))
+    return r
+end
+
+# weak normal form with leftmost word divisions
+function normal_form_weak(
+    f::FreeAssAlgElem{T},
+    g::Vector{FreeAssAlgElem{T}},
+) where {T<:FieldElement}
+    R = parent(f)
+    s = length(g)
+    while length(f) > 0
+        i = 1
+        @label again
+        ok, ml, mr = word_divides_leftmost(f.exps[1], g[i].exps[1])
+        if !ok && i < s
+            i += 1
+            @goto again
+        end
+        if ok
+            qi = divexact(f.coeffs[1], g[i].coeffs[1])
+            f = _sub_rest(f, mul_term(qi, ml, g[i], mr), 1) # enforce lt cancelation
+        else
+            break
+        end
+    end
+    return f
+end
+
+@doc raw"""
+    interreduce!(g::Vector{FreeAssAlgElem{T}})
+
+Interreduce a given Groebner basis with itself, i.e. compute the normal form of each
+element of `g` with respect to the rest of the elements and discard elements with
+normal form $0$ and duplicates.
+""" 
+function interreduce!(g::Vector{FreeAssAlgElem{T}}) where {T}
+    i = 1
+    while length(g) > 1 && length(g) >= i
+        aut = AhoCorasickAutomaton([g_j.exps[1] for g_j in g[1:end .!= i]])
+        r = normal_form(g[i], g[1:end .!= i], aut)
+        if iszero(r)
+            deleteat!(g, i)
+        elseif g[i] != r
+            g[i] = r
+            i = 1
+        else
+            i += 1
+        end
+    end
+    return g
+end
+
+## checks whether there is an overlap between a and b at position i of b
+#  such that b[i:length(b)] = a[1:length(b)-i]
+function check_left_overlap(a::Vector{Int}, b::Vector{Int}, i::Int)
+    for j in 0:(length(b) - i)
+        if j >= length(a)
+            return false # this is a center overlap
+        end
+        if b[i + j] != a[j + 1]
+            return false
+        end
+    end
+    return true
+end
+
+###
+# find all non-trivial left-obstructions of a and b
+# i.e. all words w_1 and w_2^' s.t. w_1 a = b w_2^'
+# where length(w_1) < length(b) and length(w_2^') < length(a)
+# the return vector is of the form [(w_1, w_2^'), ...]
+# if w_1 or w_2^' is empty, the corresponding obstruction is not returned
+function left_obstructions(a::Vector{Int}, b::Vector{Int})
+    v = Tuple{Vector{Int},Vector{Int}}[]
+    for i in 2:length(b)
+        if check_left_overlap(a, b, i)
+            if length(b) - i + 2 <= length(a) # w_2^' should not be empty!
+                push!(v, (b[1:(i - 1)], a[(length(b) - i + 2):length(a)]))
+            end
+        end
+    end
+    return v
+end
+
+
+###
+# find all non-trivial right-obstructions of a and b
+# i.e. all words w_2 and w_1^' s.t. a w_1^' = w_2 b
+# where length(w_1^') < length(b) and length(w_2) < length(a)
+# the return vector is of the form [(w_2, w_1^'), ...]
+# if w_1^' or w_2 is empty, the corresponding obstruction is not returned
+function right_obstructions(a::Vector{Int}, b::Vector{Int})
+    return left_obstructions(b, a)
```

Comment:
(w,v) is a left obstruction of (a,b) :<=>  (w,v) such that w a = b v
(w,v) is a right obstruction of (a,b) :<=>  (w,v) such that a w = v b

Thus:
(w,v) is a right obstructions of (b,a)
<=> (w,v) such that b w = v a
<=> (w,v) such that v a = b w
<=> (v,w) is a left obstruction of

Where is my mistake then?

Reply:
It seems to me you are swapping them in your code, too?
```julia
    for x in left_obstructions(a, b)
        push!(res, (x[1], one, one, x[2]))
    end
    for x in left_obstructions(b, a)
        push!(res, (one, x[2], x[1], one))
    end
```
At least I assume the second loop is meant to go over the right obstructions.

Reply:
I'm sorry, I misunderstood your first comment.
You are right, however in the comment above the function it is explained, that the result is to be interpreted in the reverse order.
Apparently when implementing this, we thought this would be the easiest way, however this is really quite confusing if one does not pay close attention.
I could change this or I could also remove the entire function for now, since it is currently not used anywhere.

Reply:
I find it extremely confusing that the two functions have reverse output order. Yes, it is documented, but what would you say if I provide  function `cos(x::Float)` with documentation saying "Returns the sine of `x`. To get the cosine of `x` just compute `cos(x - pi)`". Sure, it is documented, and also more convenient to implement if you already have `sin(x)`, but it is nevertheless surprising and guaranteed to cause programming errors. So I wouldn't do it.